### PR TITLE
feat(remote): bidirectional parity assertions + eager footer-indexed backward discovery

### PR DESCRIPTION
**Pull request by sametd**
## Summary

Two interlocking remote-access improvements landed in one PR because they share fixtures and the same harness extension surface:

1. **Eager footer-indexed backward discovery** (Rust sync + async + TypeScript) — when the bidirectional walker discovers a footer-indexed message via its postamble, the dispatcher folds an eager footer-region fetch into the same paired round as the candidate-preamble validation. The parsed footer's `metadata` and `index` frames land in the cached layout inline, so a subsequent `read_message_metadata(idx)` short-circuits without issuing a separate footer-region GET. Best-effort: footer fetch failure never poisons preamble validation.

2. **Parity-harness bidirectional event coverage** (`tests/remote-parity/`) — extends the cross-language HTTP parity harness with role-aware classification (`fwd_preamble`, `bwd_postamble`, `bwd_preamble_validation`), a bidirectional `RoundBuilder`, two new footer-indexed fixtures generated via `StreamingEncoder.finish_backfilled` (newly exposed in the Python binding), and a structural eager-footer assertion that fires across both languages and every fixture.

### Public API
- `tensogram::remote_scan_parse::BackwardOutcome::NeedPreambleValidation` gains `first_footer_offset: u64`.
- New `tensogram::footer_region_present(first_footer_offset, length) -> bool` predicate (checked-arithmetic, never panics).
- WASM JS shape: `{ kind: "NeedPreambleValidation", msgStart, length, firstFooterOffset }` (bigint).
- TypeScript `BackwardOutcome` discriminated union extended in lockstep.
- Python `StreamingEncoder.finish_backfilled() -> bytes` exposes the existing core `finish_with_backfill` for backward-locatable streaming output.

### Behaviour
- Bidirectional walker on footer-indexed messages issues exactly one footer-region fetch per backward-discovered message, populating metadata + index inline. Header-indexed messages stay lazy (would be 3 GETs vs forward's 1, net-worse).
- Header-indexed messages with footer hash frames have a non-empty footer region; the dispatcher fetches the bytes speculatively but discards them after the preamble flags reveal `HEADER_INDEX`. Cost: a few hundred bytes; benefit: zero extra GETs.
- Footer fetch failure (transport, parse, abort) silently falls through to the lazy path. User abort during footer fetch correctly propagates.

### Tests
- Pure parser: `footer_region_present` boundary cases (empty footer, before-preamble, valid-range, no-panic-on-underflow).
- WASM bindings: `parse_backward_postamble_outcome` JS shape pinned with `firstFooterOffset` field.
- Rust dispatcher: eager populate end-to-end on footer-indexed file, header-indexed skip, corrupt-footer fall-through.
- TypeScript dispatcher: header-indexed bidirectional walk issues no footer-region fetches; synthetic FOOTER_INDEX flag triggers eager fetch attempt.
- Python: `finish_backfilled` round-trip + dual-finish error paths + two-message concatenation pattern.
- Parity harness: new `test_eager_footer_one_fetch_per_backward_discovered_message[fixture, language]` parametrised over the two new footer-indexed fixtures × both languages, asserting `footer_region_payloads_count == floor(N/2)` (the count proves the lazy populate didn't fire, since lazy would add one more).
- Parity harness: new `TestRoundBuilderBidirectional` cases covering paired arrival shapes (fwd-then-bwd, bwd-then-fwd, role-collision, multi-round).

## Trade-offs

- **Cross-language wire-level event-parity for bidirectional mode is impossible.** Rust's `object_store.get_ranges` coalesces paired forward + backward fetches into a single multi-byte HTTP request when ranges fall under the coalesce threshold (every fixture in this harness does); TypeScript's `Promise.allSettled` issues each Range as a separate request. The per-round wire shapes therefore differ structurally. Cross-language correctness is asserted via the existing per-language layout-equality test plus the cross-language forward-only event-parity test plus the new structural eager-footer assertion that fires across both languages.
- **Async preamble + footer parallel; sync + TypeScript sequential.** Rust async uses `tokio::join!` for parallelism; Rust sync (mutex-held) and TypeScript walkers use sequential awaits, paying one extra round-trip per backward-discovered footer-indexed message in exchange for simpler abort-propagation. Acceptable; documented.
- **`bwd_preamble_validation` ScanRole reserved but not emitted.** A candidate-preamble validation fetch is structurally indistinguishable from a paired forward preamble at the wire level; the `RoundBuilder` therefore propagates the classifier-emitted role unchanged.

## Out of scope

- Default-flip benchmarks (still gating the `bidirectional=true` default across Rust + Python + TypeScript).
- Adaptive direction (deferred to post-default-flip).
- Refactoring the private + public `BackwardOutcome` parsers into one (the private one carries `Preamble`/`CachedLayout` types that can't cross the WASM boundary; both stay in lockstep, drift caught by parity-harness fixtures).

## Verification

- `cargo fmt --check` clean
- `cargo clippy --workspace --all-targets --all-features -- -D warnings` clean
- `cargo test -p tensogram --features "remote,async" --lib`: 479 lib tests pass
- `python -m pytest tests/remote-parity/ python/tests/test_streaming_encoder_backfill.py`: 202 pass
- `make ts-typecheck` clean (strict tsc)
- `make ts-test`: 395 vitest cases pass

Docs Preview
https://sites.ecmwf.int/docs/tensogram/pull-requests/PR-100

### CODE REVIEW

File: tests/remote-parity/run_parity.py [hunk 1]
```diff
@@ -45,14 +55,28 @@
...
 _FIXTURES: tuple[str, ...] = (
     "single-msg",
     "two-msg",
     "ten-msg",
     "hundred-msg",
+    "single-msg-footer",
+    "ten-msg-footer",
+)
+
+_FOOTER_FIXTURES: tuple[str, ...] = (
```
[the remark this answers is not in the document]
Comment (sametd) [reply tests/remote-parity/run_parity.py:77]: False positive — `_FOOTER_FIXTURES` is imported by `tests/remote-parity/test_parity.py` (line 42) and used at line 172 to parametrise `test_eager_footer_one_fetch_per_backward_discovered_message[fixture, language]`.  The bot inspected `run_parity.py` in isolation without checking cross-module imports.  Removing the constant would break the test.

File: typescript/tests/lazyBidirectional.test.ts [hunk 2]
```diff
@@ -485,6 +485,115 @@ describe('TensogramFile.fromUrl bidirectional walker', () => {
...
+      bidirectional: true,
+    });
+    expect(file.messageCount).toBe(4);
+
+    // Every scan-walk Range fetch is exactly PREAMBLE_BYTES (24) wide.
+    // Eager-footer fetches would be larger (footer regions span multiple
+    // CBOR frames).  On header-indexed messages the FOOTER_INDEX flag
+    // gate must skip the speculative footer fetch entirely, so the
+    // observed request log contains only 24-byte scan ranges (plus the
+    // HEAD probe).
```
[the remark this answers is not in the document]
Comment (sametd) [reply typescript/tests/lazyBidirectional.test.ts]: Fixed in 2bacf60.  The TS test comment now accurately ties the absence of footer-region fetches to the postamble-derived `footerRegionPresent(firstFooterOffset, length)` predicate that the dispatcher consults BEFORE preamble validation, instead of misattributing it to the FOOTER_INDEX flag gate (which actually only kicks in inside `tryApplyEagerFooter` after preamble validation, downstream of the fetch decision).

File: tests/remote-parity/schema.json [hunk 3]
```diff
@@ -15,15 +15,19 @@
...
+      "description": "Zero-based round index for 'scan' events. In bidirectional mode a round contains at most one forward preamble, one backward postamble, and one backward preamble validation (paired requests may arrive in either order). Non-scan events use -1."
     },
     "direction": {
       "enum": ["forward", "backward", "none"],
       "description": "Which cursor the request advances. 'none' for probes, payloads, fallbacks, and errors that are not part of a scan walk."
     },
+    "role": {
+      "enum": ["fwd_preamble", "bwd_postamble", "bwd_preamble_validation", "none"],
+      "description": "Scan-walk role assigned by the classifier and round builder. 'fwd_preamble' = forward preamble fetch at a known message start. 'bwd_postamble' = backward postamble fetch matching a known message end. 'bwd_preamble_validation' = a fwd_preamble-shaped fetch that follows a bwd_postamble in the same round. 'none' for non-scan events."
+    },
```
[the remark this answers is not in the document]
Comment (sametd) [reply tests/remote-parity/schema.json]: Fixed in 2bacf60.  Schema text now states explicitly that `bwd_preamble_validation` is reserved but not currently emitted (a candidate-preamble validation fetch is wire-shape-indistinguishable from a paired forward preamble, so the classifier propagates the original role unchanged).  `scan_round` description also rewritten to reflect the role-collision rule the simplified `RoundBuilder` actually uses.

File: rust/tensogram/src/remote.rs [hunk 4]
```diff
@@ -885,11 +900,31 @@ impl RemoteBackend {
...
+                let footer = if footer_region_present(*first_footer_offset, *length) {
+                    let footer_start = msg_start.saturating_add(*first_footer_offset);
+                    let footer_end = msg_start
+                        .saturating_add(*length)
+                        .saturating_sub(POSTAMBLE_SIZE as u64);
+                    if footer_start < footer_end {
+                        // Best-effort: silently drop on failure so the lazy
+                        // `ensure_layout` path picks up footer discovery later.
+                        self.get_range(footer_start..footer_end).ok()
+                    } else {
```
[the remark this answers is not in the document]
Comment (sametd) [reply rust/tensogram/src/remote.rs:933]: Fixed in d654b29.  `scan_bidir_round_locked`'s docstring now distinguishes three fetch tiers — required (paired fwd preamble + bwd postamble), required-when-postamble-parsed (bwd candidate preamble), and best-effort (eager footer chunk) — with explicit error/rollback semantics for each.  Format errors are documented as absorbed locally regardless of which fetch surfaces them.  The async sibling `scan_bidir_round_async` already cross-references this docstring via 'see that method's docstring' so it inherits the update transitively.
